lotman/lotman.go

```diff
@@ -0,0 +1,655 @@
+		goArr = append(goArr, goStr)
+
+		// Move to the next uintptr.
+		ptr += unsafe.Sizeof(uintptr(0))
+	}
+
+	return goArr
+}
+
+// Trim any buffer we get back from LotMan to the first null char
+func trimBuf(buf *[]byte) {
+	// Find the index of the first null character
+	nullIndex := bytes.IndexByte(*buf, 0)
+
+	// Trim the slice after the first null character
+	if nullIndex != -1 {
+		*buf = (*buf)[:nullIndex]
+	}
+}
+
+// Use the detected runtime to predict the location of the LotMan library.
+func getLotmanLib() string {
+	fallbackPaths := []string{
+		"/usr/lib64/libLotMan.so",
+		"/usr/local/lib64/libLotMan.so",
+		"/opt/local/lib64/libLotMan.so",
+	}
+
+	switch runtime.GOOS {
+	case "linux":
+		configuredPath := param.Lotman_LibLocation.GetString()
+		if configuredPath != "" {
+			if _, err := os.Stat(configuredPath); err == nil {
+				return configuredPath
+			}
+			log.Errorln("libLotMan.so not found in configured path, attempting to find using known fallbacks")
+		}
+
+		for _, path := range fallbackPaths {
+			if _, err := os.Stat(path); err == nil {
+				return path
+			}
+		}
+		panic("libLotMan.so not found in any of the known paths")
+	default:
+		panic(fmt.Errorf("GOOS=%s is not supported", runtime.GOOS))
+	}
+}
+
+func GetAuthorizedCallers(lotName string) (*[]string, error) {
+	success := InitLotman()
+	if !success {
+		return nil, errors.New("Failed to initialize LotMan")
+	}
+
+	// A caller is authorized if they own a parent of the lot. In the case of self-parenting lots, the owner is authorized.
+	errMsg := make([]byte, 2048)
+	cParents := unsafe.Pointer(nil)
+
+	// Get immediate parents (including self to determine rootliness). We'll use them to determine owners
+	// who are allowed to manipulate, and thus delete, the lot
+	ret := LotmanGetLotParents(lotName, false, true, &cParents, &errMsg)
+	if ret != 0 {
+		trimBuf(&errMsg)
+		return nil, errors.Errorf("Failed to determine %s's parents: %s", lotName, string(errMsg))
+	}
+
+	parents := cArrToGoArr(&cParents)
+
+	owners := []string{}
+	internalOwners := []string{}
+	for _, parent := range parents {
+		cOwners := unsafe.Pointer(nil)
+		LotmanGetLotOwners(parent, true, &cOwners, &errMsg)
+		if ret != 0 {
+			trimBuf(&errMsg)
+			return nil, errors.Errorf("Failed to determine appropriate owners of %s's parents: %s", lotName, string(errMsg))
+		}
+
+		internalOwners = append(internalOwners, cArrToGoArr(&cOwners)...)
+	}
+
+	// Deduplicate the owners
+	occurred := map[string]bool{}
+	for e := range internalOwners {
+		if !occurred[internalOwners[e]] {
+			occurred[internalOwners[e]] = true
+			owners = append(owners, internalOwners[e])
+		}
+	}
+
+	return &owners, nil
+}
+
+// Initialize the LotMan library and bind its functions to the global vars
+// We also perform a bit of extra setup such as setting the lotman db location
+func InitLotman() bool {
+	// If we've already tried to init the library, return the result of that attempt
+	if lotmanInitTried {
```

Comment:
Why would InitLotman be called multiple times? I wonder if Lotman is enabled, InitLotman should be called once at startup.

Reply:
Most of the linking stuff here was based on choices I made when integrating Lotman with HTCondor (integration efforts ongoing). I guess there's probably no immediate reason we'd end up calling a Lotman function if we don't initialize in the cache. I'll remove the checks.